### DISCUSSION TITLE
Lesson 7: Foundry Fund Me --Q: vm.txGasPrice is just for gasUsed view？

### DISCUSSION THREAD

**Discussion by solnex** [Cyfrin/foundry-full-course-cu#205] [Q&A]
I want run the test with gas cost，but vm.txGasPrice seems does not work like I expect.Here is the code
```
function testWithdrawFromASingleFunderWithGas() public funded {
        // Arrange
        uint256 startingFundMeBalance = address(fundMe).balance;
        uint256 startingOwnerBalance = fundMe.getOwner().balance;
        console.log(startingFundMeBalance);
        console.log(startingOwnerBalance);
        vm.txGasPrice(GAS_PRICE);
        uint256 gasStart = gasleft();
        // Act
        vm.startPrank(fundMe.getOwner());
        fundMe.withdraw();
        vm.stopPrank();

        uint256 gasEnd = gasleft();
        uint256 gasUsed = (gasStart - gasEnd) * tx.gasprice;
        console.log(gasUsed);

        // Assert
        uint256 endingFundMeBalance = address(fundMe).balance;
        uint256 endingOwnerBalance = fundMe.getOwner().balance;
        console.log(endingOwnerBalance);
        assertEq(endingFundMeBalance, 0);
        assertEq(
            startingFundMeBalance + startingOwnerBalance,
            endingOwnerBalance + gasUsed
        );
    }
```
here is the test report
```
Running 1 test for test/FundMeTester.t.sol:FundMeTester
[FAIL. Reason: Assertion failed.] testWithdrawFromASingleFunderWithGas() (gas: 96467)
Logs:
  100000000000000000
  79228162514264337593543950335
  10650
  79228162514364337593543950335
  Error: a == b not satisfied [uint]
        Left: 79228162514364337593543950335
       Right: 79228162514364337593543960985
```
79228162514364337593543950335 - 79228162514264337593543950335 = 100000000000000000 
It shows vm.txGasPrice(1) not work.

**Comment by DevSwayam** [comment]
I hope that this below
``` solidity 
 assertEq(
            startingFundMeBalance + startingOwnerBalance,
            endingOwnerBalance + gasUsed
        );
```
 test is failed.
Well when i was watching tutorial i remember that patrick told that for local chain transactions the gas is not included hence we able to write test without thinking about gas stuff. 

Hence when your writing above test just as terminal told you 
``` a==b not satisfied``` 
and the differnce between

#### a(startingFundMeBalance + startingOwnerBalance) and b(endingOwnerBalance + gasUsed) is 

 ```Solidity 
79228162514364337593543950335(a) - 79228162514364337593543960985(b) = -10,650
```

means it is clear that While testing foundry is not considering gas thing. and hence you got differnce of -10,650 which is exactly equal to your consoled gas value = 10650.

correct Test Would be 
```Solidity
assertEq(
            startingFundMeBalance + startingOwnerBalance,
            endingOwnerBalance 
        );
```

I Hope it is right. Still would like @PatrickAlphaC to cross check my answer.

**Reply by solnex** [reply]
"a-b = - gasUsed" is because the variable 'b' itself contains 'gasUsed'. What I mean is whether the gas price of transactions on the local Anvil test network can be adjusted using 'vm.txGasPrice', or if it is fixed at 0 and cannot be adjusted.

**Reply by DevSwayam** [reply]
Gas price is default to 0. But if you wanna calculate the gas prize for the interval of whole transaction make sure that all function calls gas prize should be included which you are not doing.
